xds/internal/xdsclient/authority.go

```diff
@@ -86,89 +98,135 @@ type authority struct {
 	//
 	// The second level map key is the resource name, with the value being the
 	// actual state of the resource.
-	resourcesMu sync.Mutex
-	resources   map[xdsresource.Type]map[string]*resourceState
-	closed      bool
+	resources map[xdsresource.Type]map[string]*resourceState
+
+	// An ordered list of xdsChannels corresponding to the list of server
+	// configurations specified for this authority in the bootstrap. The
+	// ordering specifies the order in which these channels are preferred for
+	// fallback.
+	xdsChannelConfigs []*xdsChannelWithConfig
+
+	// The current active xdsChannel. Here, active does not mean that the
+	// channel has a working connection to the server. It simply points to the
+	// channel that we are trying to work with, based on fallback logic.
+	activeXDSChannel *xdsChannelWithConfig
 }
 
-// authorityArgs is a convenience struct to wrap arguments required to create a
-// new authority. All fields here correspond directly to appropriate fields
-// stored in the authority struct.
-type authorityArgs struct {
-	// The reason for passing server config and bootstrap config separately
-	// (although the former is part of the latter) is because authorities in the
-	// bootstrap config might contain an empty server config, and in this case,
-	// the top-level server config is to be used.
-	serverCfg          *bootstrap.ServerConfig
-	bootstrapCfg       *bootstrap.Config
-	serializer         *grpcsync.CallbackSerializer
-	resourceTypeGetter func(string) xdsresource.Type
-	watchExpiryTimeout time.Duration
-	backoff            func(int) time.Duration // Backoff for ADS and LRS stream failures.
-	logger             *grpclog.PrefixLogger
+// authorityBuildOptions wraps arguments required to create a new authority.
+type authorityBuildOptions struct {
+	serverConfigs    bootstrap.ServerConfigs      // Server configs for the authority
+	name             string                       // Name of the authority
+	serializer       *grpcsync.CallbackSerializer // Callback serializer for invoking watch callbacks
+	getChannelForADS xdsChannelForADS             // Function to acquire a reference to an xdsChannel
+	logPrefix        string                       // Prefix for logging
 }
 
-func newAuthority(args authorityArgs) (*authority, error) {
+// newAuthority creates a new authority instance with the provided
+// configuration. The authority is responsible for managing the state of
+// resources requested from the management server, as well as acquiring and
+// releasing references to channels used to communicate with the management
+// server.
+//
+// Note that no channels to management servers are created at this time. Instead
+// a channel to the first server configuration is created when the first watch
+// is registered, and more channels are created as needed by the fallback logic.
+func newAuthority(args authorityBuildOptions) *authority {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Not applicable to this PR; but just to remind about context scoping discussion for the future when you refactor xDS Client @purnesh42H